### PR TITLE
Move import check to before state reset

### LINKED ISSUE #23898: NameError: name 'AcceleratorState' is not defined
### System Info

This script fails on 68d53bc7178866821282f45732c1e465f5160fa6 but passes on de9255de27abfcae4a1f816b904915f0b1e23cd9

Hopefully the problem is pretty clear from the message.

```
(/home/ezyang/local/debug/pytorch-env) [ezyang@devgpu019.ftw1 ~/local/debug]$ pp python transformers/examples/pytorch/speech-recognition/run_speech_recognition_ctc.py     --dataset_name="common_voice"     --model_name_or_path="facebook/wav2vec2-large-xlsr-53"     --dataset_config_name="tr"     --output_dir="./wav2vec2-common_voice-tr-demo-dist"     --preprocessing_num_workers="16"     --overwrite_output_dir     --num_train_epochs="15"     --per_device_train_batch_size="4"     --gradient_accumulation_steps="1"     --learning_rate="3e-4"     --warmup_steps="500"     --evaluation_strategy="steps"     --text_column_name="sentence"     --save_steps="400"     --eval_steps="100"     --logging_steps="1"     --layerdrop="0.0"     --save_total_limit="3"     --freeze_feature_encoder     --gradient_checkpointing     --chars_to_ignore , ? . ! - \; \: \" “ % ‘ ” �     --fp16     --group_by_length     --do_train --do_eval     --torch_compile True
Traceback (most recent call last):
  File "/data/users/ezyang/debug/transformers/examples/pytorch/speech-recognition/run_speech_recognition_ctc.py", line 775, in <module>
    main()
  File "/data/users/ezyang/debug/transformers/examples/pytorch/speech-recognition/run_speech_recognition_ctc.py", line 380, in main
    model_args, data_args, training_args = parser.parse_args_into_dataclasses()
  File "/home/ezyang/local/debug/pytorch-env/lib/python3.10/site-packages/transformers/hf_argparser.py", line 346, in parse_args_into_dataclasses
    obj = dtype(**inputs)
  File "<string>", line 111, in __init__
  File "/home/ezyang/local/debug/pytorch-env/lib/python3.10/site-packages/transformers/training_args.py", line 1340, in __post_init__
    and (self.device.type != "cuda")
  File "/home/ezyang/local/debug/pytorch-env/lib/python3.10/site-packages/transformers/training_args.py", line 1764, in device
    return self._setup_devices
  File "/home/ezyang/local/debug/pytorch-env/lib/python3.10/site-packages/transformers/utils/generic.py", line 54, in __get__
    cached = self.fget(obj)
  File "/home/ezyang/local/debug/pytorch-env/lib/python3.10/site-packages/transformers/training_args.py", line 1670, in _setup_devices
    AcceleratorState._reset_state()
NameError: name 'AcceleratorState' is not defined
```

### Who can help?

_No response_

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

`python transformers/examples/pytorch/speech-recognition/run_speech_recognition_ctc.py     --dataset_name="common_voice"     --model_name_or_path="facebook/wav2vec2-large-xlsr-53"     --dataset_config_name="tr"     --output_dir="./wav2vec2-common_voice-tr-demo-dist"     --preprocessing_num_workers="16"     --overwrite_output_dir     --num_train_epochs="15"     --per_device_train_batch_size="4"     --gradient_accumulation_steps="1"     --learning_rate="3e-4"     --warmup_steps="500"     --evaluation_strategy="steps"     --text_column_name="sentence"     --save_steps="400"     --eval_steps="100"     --logging_steps="1"     --layerdrop="0.0"     --save_total_limit="3"     --freeze_feature_encoder     --gradient_checkpointing     --chars_to_ignore , ? . ! - \; \: \" “ % ‘ ” �     --fp16     --group_by_length     --do_train --do_eval     --torch_compile True`

### Expected behavior

doesn't error this way

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR moves the reset of the `AcceleratorState` to be under the check for if it is available, and guards the resetting better.

Fixes # (issue)

Fixes https://github.com/huggingface/transformers/issues/23898

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger

### CODE REVIEW

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -1667,12 +1667,12 @@ def ddp_timeout_delta(self) -> timedelta:
...
-        if not is_sagemaker_mp_enabled() and not is_accelerate_available(check_partial_state=True):
-            raise ImportError(
-                "Using the `Trainer` with `PyTorch` requires `accelerate>=0.19.0`: Please run `pip install transformers[torch]` or `pip install accelerate -U`"
-            )
+        if not is_sagemaker_mp_enabled():
+            if not is_accelerate_available(check_partial_state=True):
+                raise ImportError(
+                    "Using the `Trainer` with `PyTorch` requires `accelerate>=0.19.0`: Please run `pip install transformers[torch]` or `pip install accelerate -U`"
+                )
+            AcceleratorState._reset_state(reset_partial_state=True)
```
Comment (sgugger) [inline src/transformers/training_args.py:1675]: Was the `PartialState._reset_state` line unnecessary?

**Review by sgugger — APPROVED** [review]
Thanks!

File: src/transformers/training_args.py (hunk 1 above)
Comment (muellerzr) [reply src/transformers/training_args.py:1675]: Yes, `AcceleratorState` can reset the partial state for us too, along with its own state :)

**Comment by pacman100** [comment]
How about adding accelerate as a required dependency in the setup.py?

**Comment by sgugger** [comment]
torch is not a hard dependency of Transformers.

**Comment by muellerzr** [comment]
Which is why it's part of the `[torch]` reqs :)

**Pull request merged**